File: programmers/algorithm_data_structure/greedy/단속카메라.py
def my_solution(routes):
    '''
    * 아래 best 풀이 반드시 참고 *
    => 훨씬 간단하고 더 greedy하며 빠르게 풀어냈다.

    차의 동시 만족 범위를 최대한 겹치어
    범위가 1 미만이 되었을 때 단속 카메라 설치 후
    다시 범위를 계산한다
    '''
    answer = 0
    sorted_routes = sorted(routes, reverse=True)

    last_in, first_out = sorted_routes.pop()
    while sorted_routes:
        car_in, car_out = sorted_routes.pop()
        last_in = (car_in if car_in > last_in else last_in)
        first_out = (car_out if car_out < first_out else first_out)

        if last_in > first_out:
            answer += 1
            last_in = car_in
            first_out = car_out

    return answer + 1  # 나머지 처리


def best_solution(routes):
    '''
    마지막으로 설치된 카메라 지점이 
    차의 진입 지점 이전에 있으면
    새로운 카메라를 차의 진입 이후 지점에 설치한다.
    (최대한 다른 차도 수용하는 local optimal)
    '''
    routes = sorted(routes, key=lambda x: x[1])
    last_camera = -30000

    answer = 0

    for route in routes:
        if last_camera < route[0]:
            answer += 1
            last_camera = route[1]

    return answer

File: programmers/algorithm_data_structure/greedy/단속카메라.java
package algorithm_data_structure.greedy;

import java.util.Arrays;
import java.util.LinkedList;

class 단속카메라 {
    /*
     * 자동차가 나가는 순서로 보았을 때
     * 아직 카메라 배치가 안된 다른 자동차는 지금 자동차보다 늦게 나가고
     * 지금 자동차보다 일찍 들어오거나 늦게 들어온다.
     * 
     * 이때 지금 자동차 포함하여 가장 많이 자동차 범위가 겹치는 구간은
     * 자동차가 나가는 그 순간이며,
     * 이때 도로 위에 오른 자동차들은 그 카메라 한 대로 다 처리 가능하다.
     * 
     * 이런 개념으로 카메라를 배치해나감
     * (기본적으로 routes가 들어온 순서대로 정렬되었다고 가정 + 나간 순서대로 정렬)
     */
    public int mySolution(int[][] routes) {
        int answer = 0;
        LinkedList<int[]> sortedRoutes = new LinkedList<>();

        for (int[] route : routes) {
            sortedRoutes.add(route);
        }
        sortedRoutes.sort((o1, o2) -> o1[1] - o2[1]);

        while (!sortedRoutes.isEmpty()) {
            int[] route = sortedRoutes.removeFirst();

            while(!sortedRoutes.isEmpty() && sortedRoutes.getFirst()[0] <= route[1]) {
                sortedRoutes.removeFirst();
            }
            answer += 1;
        }
 
        return answer;
    }

    /*
     * 위의 개념과 동일한 풀이지만 
     * for문과 if문으로 간단하게 풀이해냈다.
     * 
     * 위 코드보다 속도가 약간 빠르다.
     * 일단 빼고 넣는 게 없으며, 조건문 갯수도 적다.
     */
    public int bestSolution(int[][] routes) {
        int answer = 0;
        int last_camera = Integer.MIN_VALUE;
        int[][] sortedRoutes = new int[routes.length][2];

        for (int i = 0; i < routes.length; i++) {
            sortedRoutes[i] = routes[i];
        }
        Arrays.sort(sortedRoutes, (o1, o2) -> o1[1] - o2[1]);
        
        for (int[] a : sortedRoutes) {
            if (last_camera < a[0]) {
                answer += 1;
                last_camera = a[1];
            }
        }
        
        return answer;
    }
}